server/src/main/java/org/elasticsearch/action/admin/indices/shrink/TargetNumberOfShardsCalculator.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.shrink;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.IndexWriter;
+import org.elasticsearch.action.admin.indices.stats.IndexShardStats;
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsResponse;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.shard.DocsStats;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.StoreStats;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.IntFunction;
+
+/**
+ * This class calculates and verifies the number of shards of a target index after a resize operation. It supports three operations SHRINK,
+ * SPLIT and CLONE.
+ */
+public interface TargetNumberOfShardsCalculator {
+
+    /**
+     * Calculates the target number of shards based on the parameters of the request
+     * @param numberOfShards requested number of shards or null if it was not provided
+     * @param maxPrimaryShardSize requested max primary shard size or null if it was not provided
+     * @param sourceMetadata the index metadata of the source index
+     * @return the number of shards for the target index
+     */
+    int calculate(@Nullable Integer numberOfShards, ByteSizeValue maxPrimaryShardSize, IndexMetadata sourceMetadata);
```

Comment:
I believe the `maxPrimaryShardSize` can be nullable too. Shall we annotate it?

Reply:
Definitely, I missed it